lib/db_utils.go

```diff
@@ -6273,9 +6273,75 @@ func DBGetLowestDAOCoinAskOrders(txn *badger.Txn, inputOrder *DAOCoinLimitOrderE
 	return orders, nil
 }
 
-func DBGetHighestDAOCoinBidOrder(txn *badger.Txn, inputOrder *DAOCoinLimitOrderEntry) (*DAOCoinLimitOrderEntry, error) {
-	// TODO
-	return nil, nil
+func DBGetHighestDAOCoinBidOrders(txn *badger.Txn, inputOrder *DAOCoinLimitOrderEntry, startKey []byte) ([]*DAOCoinLimitOrderEntry, error) {
+	queryOrder := inputOrder.Copy()
+	requestedQuantity := queryOrder.Quantity
+
+	// Confirm that the input order is an ask order.
+	if queryOrder.OperationType != DAOCoinLimitOrderEntryOrderTypeAsk {
+		return nil, fmt.Errorf("DBGetHighestDAOCoinBidOrder: input must be an ask order")
+	}
+
+	// Convert the input ask order to the bid-order params to query for.
+	// Retain the input ask order's:
+	//   * DenominatedCoinType
+	//   * DenominatedCoinCreatorPKID
+	//   * DAOCoinCreatorPKID
+	// Convert:
+	//   * OperationType from Ask to Bid
+	//   * PriceNanos to MAX_UINT256
+	//   * BlockHeight to MAX_UINT32
+	//   * Quantity to MAX_UINT256
+	queryOrder.OperationType = DAOCoinLimitOrderEntryOrderTypeBid
+	queryOrder.PriceNanos = *MaxUint256.Clone()
+	queryOrder.BlockHeight = math.MaxUint32
+	queryOrder.Quantity = *MaxUint256.Clone()
+
+	key := DBKeyForDAOCoinLimitOrder(queryOrder, false)
+
+	// If passed a start key, start seeking from there.
+	if startKey != nil {
+		key = startKey
+	}
+
+	opts := badger.DefaultIteratorOptions
+	opts.Reverse = true // Go in reverse order since a higher bid price is better.
+	iterator := txn.NewIterator(opts)
+	defer iterator.Close()
+
+	// Seek first matching order.
+	orders := []*DAOCoinLimitOrderEntry{}
+
+	for iterator.Seek(key); iterator.ValidForPrefix(key) && requestedQuantity.GtUint64(0); iterator.Next() {
+		// If picking up from where you left off, skip the first order which has already been included.
+		if reflect.DeepEqual(key, startKey) {
```

Comment:
reflection.DeepEqual is a mildly expensive operation. To speed this up, let's skip this check if:
1. startKey is nil
2. we've already seen the start key - add a boolean to track this.